halo2_proofs/src/plonk/lookup/prover.rs

```diff
@@ -397,23 +386,32 @@ impl<'a, C: CurveAffine, Ev: Copy + Send + Sync + 'a> Committed<C, Ev> {
         l0: poly::AstLeaf<Ev, ExtendedLagrangeCoeff>,
         l_blind: poly::AstLeaf<Ev, ExtendedLagrangeCoeff>,
         l_last: poly::AstLeaf<Ev, ExtendedLagrangeCoeff>,
+        domain: &EvaluationDomain<C::Scalar>,
+        coset_evaluator: &mut poly::Evaluator<Ev, C::Scalar, ExtendedLagrangeCoeff>,
     ) -> (
         Constructed<C>,
         impl Iterator<Item = poly::Ast<Ev, C::Scalar, ExtendedLagrangeCoeff>> + 'a,
+        Vec<poly::AstLeaf<Ev, ExtendedLagrangeCoeff>>,
     ) {
         let permuted = self.permuted;
 
         let active_rows = poly::Ast::one() - (poly::Ast::from(l_last) + l_blind);
         let beta = poly::Ast::ConstantTerm(*beta);
         let gamma = poly::Ast::ConstantTerm(*gamma);
 
+        let product_coset =
+            coset_evaluator.register_poly(domain.coeff_to_extended(self.product_poly.clone()));
+        let permuted_input_coset = coset_evaluator
+            .register_poly(domain.coeff_to_extended(permuted.permuted_input_poly.clone()));
+        let permuted_table_coset = coset_evaluator
+            .register_poly(domain.coeff_to_extended(permuted.permuted_table_poly.clone()));
```

Comment:
How hard would it be with your approach to generate all these polynomials just in time for when they are needed in the lookup calculations (and so afterwards they can be immediately thrown away again)? These take up a lot of memory because they are in the extended domain.

I kind of think because of reasons like this where the evaluator needs special logic to more efficiently handle cases like this that both approaches will kind of evolve to largely the same thing, where some calculations need to be dynamic, but when things can be done more efficiently because code/data dependencies always remain the same it makes sense to just do that directly in code.

Reply:
do you mean fully automatically "generate on-need; and release after use"?    It may need a lot of rewrite of Evaluator, like add some GC mechanism.

Currently, the coset polys used in lookup proving are generated and released [within these two lines](https://github.com/appliedzkp/halo2/blob/9c9662bff61fbae2dc2934e52966dbae9763ec3e/halo2_proofs/src/plonk/prover.rs#L628-L640).   So it is not fully automatic, but can be done efficiently if we arrange codes carefully

Reply:
Ah I see, then in that respect this is very similar to my approach already. I do think that if more optimizations are implemented with your approach things will look similar to the code in my PR (the code there is actually decent now so easier to compare and understand I believe).